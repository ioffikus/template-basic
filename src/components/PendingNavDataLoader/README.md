
Компонент PendingNavDataLoader пока выполняет три функции:

1. как следует из его названия, ожидание загрузки данных при навигации
2. выставление флагов abortToken для прекращения загрузки неактуальных данных
3. выставление флага isLoading в ducks/progress для отображения Spin

Все эти функции компонент выполняет только в режиме SPA, т.к. в режиме SSR такое не требуется. 

Как это работает?

При изменении location, в lifecycle-методе componentWillReceiveProps выставляется флаг isPreviousLocation, который запрещает обновление компонента в lifecycle-методе shouldComponentUpdate (но когда новые данные будут загружены, выключение этого флага инициирует обновление компонента).

Функция matchRoutes из react-router-config проходит по дереву src/routes и возвращает ветку из вложенных компонентов для роута нового location. Далее для каждого такого компонента вызывается статический асинхронный метод fetchData (статический - потому что компоненты еще не инстанциированы, асинхронный - возвращает Promise по соглашению) с параметром abortToken, о котором будет сказана ниже. 

Метод fetchData имеет смысл реализовывать только для компонентов, которые участвуют в src/routes; для остальных компонентов метов fetchData никогда не будет вызван.

Метод fetchData вызывается аналогичным образом в режиме SSR (только с пустым параметром abortToken, т.к. механизм отмены запросов к API не используется в режиме SSR) внутри src/server/routes/react, благодаря чему мы имеем единообразный способ начальной загрузки данных как для SSR, так и для SPA. 

Импорт объекта abortTokenPool из src/core/API нужен, чтобы складывать туда abortToken-ы, по которым требуется отменить запросы к API - реализовано на перехватчиках (axios.interceptors.request и axios.interceptors.response).

В компоненте PendingNavDataLoader есть поле pendingAbortToken. Если пользователь выполнит переход на новую страницу, не дождавшись загрузки данных из API для текущей страницы, то pendingAbortToken попадает в abortTokenPool, чтобы отменить загрузку данных из API для ожидаемого abortToken.

Идея заимствована тут: https://www.npmjs.com/package/react-router-config
