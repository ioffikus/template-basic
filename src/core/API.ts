import axios, { AxiosRequestConfig } from 'axios';
import { IXShopHeaders } from 'src/core/interfaces/IXShopHeaders';
import ServerCfg from 'src/configs/server';

export const abortTokenPool: { [abortToken: string]: boolean } = {};

axios.interceptors.request.use(
  config => {
    if (process.env.BROWSER) {
      const abortToken = config.headers['x-abort-token'];
      if (abortToken && abortTokenPool[abortToken]) {
        return Promise.reject(null);
      }
    }
    return config;
  },
  error => {
    return Promise.reject(error);
  },
);

axios.interceptors.response.use(
  response => {
    if (process.env.BROWSER) {
      const abortToken = response.config.headers['x-abort-token'];
      if (abortToken && abortTokenPool[abortToken]) {
        return Promise.reject(null);
      }
    }
    return response;
  },
  error => {
    return Promise.reject(error);
  },
);

const _request = (method: string, config: AxiosRequestConfig, xShopHeaders: IXShopHeaders, abortToken: string) => {
  const headers =
    process.env.BROWSER && abortToken
      ? { ...config.headers, ...xShopHeaders, 'x-abort-token': abortToken }
      : { ...config.headers, ...xShopHeaders };
  const baseURL = process.env.BROWSER ? '' : `http://${ServerCfg.HOST}:${ServerCfg.PORT}`;
  return axios({
    responseType: 'json',
    validateStatus: status => status === 200,
    ...config,
    method,
    headers,
    baseURL,
  });
};

export const GET = (config: AxiosRequestConfig, xShopHeaders: IXShopHeaders, abortToken?: string) =>
  _request('GET', config, xShopHeaders, abortToken);

export const POST = (config: AxiosRequestConfig, xShopHeaders: IXShopHeaders, abortToken?: string) =>
  _request('POST', config, xShopHeaders, abortToken);

export const DELETE = (config: AxiosRequestConfig, xShopHeaders: IXShopHeaders, abortToken?: string) =>
  _request('DELETE', config, xShopHeaders, abortToken);

export const PUT = (config: AxiosRequestConfig, xShopHeaders: IXShopHeaders, abortToken?: string) =>
  _request('PUT', config, xShopHeaders, abortToken);
