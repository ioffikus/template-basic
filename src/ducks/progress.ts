import { createAction, createReducer } from 'redux-act';

export const REDUCER = 'PROGRESS';
const NS = `${REDUCER}__`;

export interface IProgressState {
  isLoading: boolean;
  fillLoad: boolean;
}

export const initialState: IProgressState = {
  isLoading: false,
  fillLoad: false,
};

const reducer = createReducer<IProgressState>({}, initialState);

export const showProgress = createAction(`${NS}SHOW_PROGRESS`);
reducer.on(showProgress, state => ({
  ...state,
  isLoading: true,
}));

export const fillProgress = createAction(`${NS}FILL_PROGRESS`);
reducer.on(fillProgress, state => ({
  ...state,
  fillLoad: true,
}));

export const hideProgress = createAction(`${NS}HIDE_PROGRESS`);
reducer.on(hideProgress, state => ({
  ...state,
  isLoading: false,
  fillLoad: false,
}));

export default reducer;
