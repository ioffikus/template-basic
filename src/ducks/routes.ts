import { createAction, createReducer } from 'redux-act';

export const REDUCER = 'ROUTES';
const NS = `${REDUCER}__`;

export interface IRoutesState extends Alicanto.Models.Routes {}

export const initialState: IRoutesState = {};

const reducer = createReducer<IRoutesState>({}, initialState);

export const setRoutes = createAction<Alicanto.Models.Routes>(`${NS}SET_ROUTES`);
reducer.on(setRoutes, (_, routes) => ({
  ...routes,
}));

export default reducer;
