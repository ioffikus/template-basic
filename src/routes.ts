import { RouteConfig } from 'react-router-config';
import AppLayout from 'src/components/AppLayout';
import CatalogLayout from 'src/components/CatalogLayout';
import HomeLayout from 'src/components/HomeLayout';
import Cart from 'src/components/Cart';
import Login from 'src/components/Login';
import NotFound, { RedirectTo404 } from 'src/components/NotFound';
import Page from 'src/components/Page';
import Product from 'src/components/Product';
import User from 'src/components/User';

const routes: RouteConfig[] = [
  {
    path: '/404',
    component: NotFound,
  },
  {
    component: AppLayout,
    routes: [
      {
        path: '/',
        exact: true,
        component: HomeLayout,
      },
      {
        path: '/catalog*', // FIXED: use `*`, matchRoutes for SSR does not work with /catalog?search=xxx, only with /catalog/?search=xxx
        component: CatalogLayout,
      },
      {
        path: '/login',
        component: Login,
      },
      {
        path: '/product/:uid',
        component: Product,
      },
      {
        path: '/cart',
        component: Cart,
      },
      {
        path: '/pages/:slug',
        component: Page,
      },
      {
        path: '/user',
        exact: true,
        component: User,
      },
      {
        path: '/user/:key',
        component: User,
      },
      {
        path: '*',
        component: RedirectTo404,
      },
    ],
  },
];

export default routes;
