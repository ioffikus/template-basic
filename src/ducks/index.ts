// instead this, we have router in store via connected-react-router
// import { routerReducer as routing, RouterState } from 'react-router-redux';
import { combineReducers } from 'redux';
import auth from './auth';
import cart from './cart';
import cartPaidItems from './cartPaidItems';
import catalogQuery from './catalogQuery';
import catalog from './catalog';
import categories from './categories';
import categoriesMenu from './categoriesMenu';
import filter from './filter';
import locale from './locale';
import meta from './meta';
import page from './page';
import product from './product';
import progress from './progress';
import popularItems from './popularItems';
import recommendItems from './recommendItems';
import routes from './routes';
import ui from './ui';
import user from './user';
import userDelivery from './userDelivery';
import userOrders from './userOrders';

export default combineReducers<Store.IState>({
  ui,
  auth,
  user,
  categories,
  categoriesMenu,
  catalogQuery,
  catalog,
  product,
  progress,
  routes,
  meta,
  filter,
  cart,
  cartPaidItems,
  userDelivery,
  userOrders,
  popularItems,
  recommendItems,
  page,
  locale,
});
