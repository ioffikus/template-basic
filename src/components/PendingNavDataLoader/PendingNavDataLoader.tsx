import React from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import nanoid from 'nanoid';
import { Route } from 'react-router';
import { RouterState } from 'connected-react-router';
import { RouteConfig, matchRoutes, renderRoutes } from 'react-router-config';
import dispatch from 'src/store/dispatch';
import { showProgress, hideProgress } from 'src/ducks/progress';
import { abortTokenPool } from 'src/core/API';
import { logException } from 'src/helpers/sentryClient';
import ConfigHelper from 'src/helpers/configHelper';

interface IOwnProps {
  routes: RouteConfig[];
}

interface IConnectedState {
  router: RouterState;
}

type IProps = IOwnProps & IConnectedState;

interface IState {
  isPreviousLocation: boolean;
}

const mapStateToProps = (state: Store.IState) => ({
  router: state.router,
});

const WrappedPendingNavDataLoader = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class PendingNavDataLoader extends React.Component<IProps, IState> {
    pendingAbortToken: string = null;

    state: IState = {
      isPreviousLocation: false,
    };

    renderRoute = () => {
      return renderRoutes(this.props.routes);
    };

    run = (promises: Promise<any>[], currentAbortToken: string) => {
      dispatch(showProgress());
      return Promise.all(promises)
        .catch(error => {
          if (_.isNull(error)) {
            return;
          }
          if (ConfigHelper.isDev()) {
            console.error(error);
          } else {
            logException(error);
          }
        })
        .then(() => {
          if (abortTokenPool[currentAbortToken]) {
            delete abortTokenPool[currentAbortToken];
          } else if (currentAbortToken === this.pendingAbortToken) {
            dispatch(hideProgress());
            this.pendingAbortToken = null;
            // clear previousLocation so the next screen renders
            this.setState({
              isPreviousLocation: false,
            });
          }
        });
    };

    getAllFetchData = (locationPathname: string) => {
      const { routes } = this.props;
      const branch = matchRoutes(routes, locationPathname);
      return branch.map(({ route }) => {
        const fetchData: Store.IFetchDataFn = _.get(route, 'component.fetchData', null);
        return _.isNull(fetchData) ? Promise.resolve() : fetchData(this.pendingAbortToken);
      });
    };

    componentWillReceiveProps(nextProps: IProps) {
      if (nextProps.router.location !== this.props.router.location) {
        if (_.isNull(this.pendingAbortToken)) {
          // save the location so we can render the old screen
          this.setState({
            isPreviousLocation: true,
          });
        } else {
          // some more changes for location
          abortTokenPool[this.pendingAbortToken] = true;
        }
        this.pendingAbortToken = nanoid();
        // load data while the old screen remains
        const promises = this.getAllFetchData(nextProps.router.location.pathname);
        this.run(promises, this.pendingAbortToken);
      }
    }

    shouldComponentUpdate(nextProps: IProps, nextState: IState) {
      if (nextState.isPreviousLocation) {
        return false;
      }
      return true;
    }

    render() {
      const location = { ...this.props.router.location, state: {} };
      return <Route location={location} render={this.renderRoute} />;
    }
  },
);

export default WrappedPendingNavDataLoader;
