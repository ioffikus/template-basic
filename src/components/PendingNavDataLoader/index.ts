import PendingNavDataLoader from './PendingNavDataLoader';

export default PendingNavDataLoader;
