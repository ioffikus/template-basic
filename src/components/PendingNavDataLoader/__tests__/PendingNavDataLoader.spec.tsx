import React from 'react';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import nanoid from 'nanoid';
import routes from 'src/routes';
import { abortTokenPool } from 'src/core/API';
import { setupDispatch } from 'src/store/dispatch';
import { initialState as progressInitialState, showProgress, hideProgress } from 'src/ducks/progress';
import routerStateMock from 'src/store/__mocks__/routerStore';
import PendingNavDataLoader from '../PendingNavDataLoader';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('src/helpers/configHelper', () => {
  return {
    isDev: jest.fn(() => {
      return process.env.NODE_ENV_FAKE === 'development';
    }),
    getEnvConstants: jest.fn(() => {
      return {};
    }),
    getEnv: jest.fn(() => {
      return process.env.NODE_ENV_FAKE;
    }),
  };
});
jest.mock('src/ducks/progress', () => {
  return {
    showProgress: jest.fn(() => {
      return () => Promise.resolve();
    }),
    hideProgress: jest.fn(() => {
      return () => Promise.resolve();
    }),
  };
});

describe(null, () => {
  process.env.NODE_ENV_FAKE = 'development';
  it('render()', () => {
    const store = mockStore({ router: routerStateMock, progress: progressInitialState });
    setupDispatch({ store });
    const wrapper = shallow(<PendingNavDataLoader routes={routes} />, { context: { store } });
    expect(wrapper.dive()).toMatchSnapshot();
  });
  it('shouldComponentUpdate() with (& without) isPreviousLocation', () => {
    const store = mockStore({ router: routerStateMock, progress: progressInitialState });
    setupDispatch({ store });
    const wrapper = shallow(<PendingNavDataLoader routes={routes} />, { context: { store } });
    const dummyWrapper = wrapper.dive();
    const dummyInstance = dummyWrapper.instance();
    expect(dummyInstance.shouldComponentUpdate({}, { isPreviousLocation: true }, {})).toBeFalsy();
    expect(dummyInstance.shouldComponentUpdate({}, { isPreviousLocation: false }, {})).toBeTruthy();
  });
  it('getAllFetchData() must return at least one promise in array', () => {
    const store = mockStore({ router: routerStateMock, progress: progressInitialState });
    setupDispatch({ store });
    const wrapper = shallow(<PendingNavDataLoader routes={routes} />, { context: { store } });
    const dummyWrapper = wrapper.dive();
    const dummyInstance = dummyWrapper.instance() as any;
    dummyInstance.pendingAbortToken = nanoid();
    const promises = dummyInstance.getAllFetchData('/');
    expect(promises[0]).toEqual(Promise.resolve(promises[0]));
  });
  it('run() when currentAbortToken in abortTokenPool', () => {
    const store = mockStore({ router: routerStateMock, progress: progressInitialState });
    setupDispatch({ store });
    const wrapper = shallow(<PendingNavDataLoader routes={routes} />, { context: { store } });
    const dummyWrapper = wrapper.dive();
    const dummyInstance = dummyWrapper.instance() as any;
    const promisesMock = [Promise.resolve()];
    const currentAbortToken = nanoid();
    abortTokenPool[currentAbortToken] = true;
    return dummyInstance.run(promisesMock, currentAbortToken).then(() => {
      expect(showProgress).toBeCalled();
      expect(abortTokenPool[currentAbortToken]).toBeUndefined();
    });
  });
  it('run() when currentAbortToken == pendingAbortToken', () => {
    const store = mockStore({ router: routerStateMock, progress: progressInitialState });
    setupDispatch({ store });
    const wrapper = shallow(<PendingNavDataLoader routes={routes} />, { context: { store } });
    const dummyWrapper = wrapper.dive();
    const dummyInstance = dummyWrapper.instance() as any;
    dummyInstance.setState({ isPreviousLocation: true });
    const promisesMock = [Promise.resolve()];
    const currentAbortToken = nanoid();
    dummyInstance.pendingAbortToken = currentAbortToken;
    return dummyInstance.run(promisesMock, currentAbortToken).then(() => {
      expect(showProgress).toBeCalled();
      expect(hideProgress).toBeCalled();
      expect(dummyInstance.pendingAbortToken).toBeNull();
      expect(dummyInstance.state).toEqual({ isPreviousLocation: false });
    });
  });
  it('change location with pendingAbortToken', () => {
    const store = mockStore({ router: routerStateMock, progress: progressInitialState });
    setupDispatch({ store });
    const wrapper = shallow(<PendingNavDataLoader routes={routes} />, { context: { store } });
    const dummyWrapper = wrapper.dive();
    const dummyInstance = dummyWrapper.instance() as any;
    const pendingAbortToken = nanoid();
    dummyInstance.pendingAbortToken = pendingAbortToken;
    dummyWrapper.setProps({ router: { ...routerStateMock, location: { pathname: '/404' } } });
    expect(abortTokenPool[pendingAbortToken]).toBeTruthy();
    delete abortTokenPool[pendingAbortToken];
  });
  it('change location without pendingAbortToken', () => {
    const store = mockStore({ router: routerStateMock, progress: progressInitialState });
    setupDispatch({ store });
    const wrapper = shallow(<PendingNavDataLoader routes={routes} />, { context: { store } });
    const dummyWrapper = wrapper.dive();
    const dummyInstance = dummyWrapper.instance();
    dummyWrapper.setProps({ router: { ...routerStateMock, location: { pathname: '/404' } } });
    expect(dummyInstance.state).toEqual({ isPreviousLocation: true });
  });
  delete process.env.NODE_ENV_FAKE;
});
