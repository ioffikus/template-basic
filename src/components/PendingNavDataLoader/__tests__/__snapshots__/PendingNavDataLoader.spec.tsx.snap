// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render() 1`] = `
<Route
  location={
    Object {
      "hash": "",
      "key": "dummy",
      "pathname": "/",
      "search": "",
      "state": Object {},
    }
  }
  render={[Function]}
/>
`;
